samples/go/downloader/downloader.go

```diff
@@ -0,0 +1,326 @@
+// Copyright 2016 Attic Labs, Inc. All rights reserved.
+// Licensed under the Apache License, version 2.0:
+// http://www.apache.org/licenses/LICENSE-2.0
+
+package main
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"os"
+	"sync"
+
+	"github.com/attic-labs/noms/go/config"
+	"github.com/attic-labs/noms/go/d"
+	"github.com/attic-labs/noms/go/datas"
+	"github.com/attic-labs/noms/go/diff"
+	"github.com/attic-labs/noms/go/marshal"
+	"github.com/attic-labs/noms/go/spec"
+	"github.com/attic-labs/noms/go/types"
+	"github.com/attic-labs/noms/go/util/exit"
+	"github.com/attic-labs/noms/go/util/status"
+	"github.com/attic-labs/noms/go/util/verbose"
+	flag "github.com/juju/gnuflag"
+)
+
+var (
+	lastSourcePathFieldName = "sourcePath"
+	foundInCacheCnt         = &counter{}
+)
+
+func main() {
+	if !download() {
+		exit.Fail()
+	}
+}
+
+func usage() {
+	fmt.Println("usage: downloader --in-path <pathspec> --out-ds <dsname> --cache-ds <dsname>")
+	flag.PrintDefaults()
+}
+
+type RemoteResource struct {
+	Url string
+}
+
+type LocalResource struct {
+	Url        string
+	Downloaded bool
+	Err        string    `noms:",omitempty"`
+	BlobRef    types.Ref `noms:",omitempty"`
+}
+
+func download() (win bool) {
+	var inPathArg = flag.String("in-path", "", "root of object hierarchy to scan for RemotePhoto objects")
+	var outDsArg = flag.String("out-ds", "", "name of output dataset")
+	var cacheDsArg = flag.String("cache-ds", "", "name of photo-cache dataset")
```

Comment:
Having a separate cache-ds will be less compatible with a future event loop world. Consider using just one dataset now?